src/query/storages/fuse/src/operations/commit.rs

```diff
@@ -393,6 +395,59 @@ impl FuseTable {
                 warn!("write last snapshot hint failure. {}", e);
             })
     }
+
+    pub async fn commit_mutation(
+        &self,
+        ctx: Arc<dyn TableContext>,
+        base_snapshot: Arc<TableSnapshot>,
+        mut segments: Vec<Location>,
+        mut summary: Statistics,
+    ) -> Result<()> {
+        let snapshot_opt = self.read_table_snapshot(ctx.clone()).await?;
+        let latest_snapshot = snapshot_opt.ok_or_else(|| {
+            ErrorCode::UnknownException("Data mutates during operation".to_string())
+        })?;
+        if latest_snapshot.snapshot_id != base_snapshot.snapshot_id {
+            if latest_snapshot.segments.len() < base_snapshot.segments.len() {
+                return Err(ErrorCode::UnknownException(
+                    "Data mutates during operation".to_string(),
+                ));
+            }
+
+            // Check if there is only insertion during the operation.
+            let mut new_segments = latest_snapshot.segments.clone();
+            let suffix = new_segments
+                .split_off(latest_snapshot.segments.len() - base_snapshot.segments.len());
+            if suffix.ne(&base_snapshot.segments) {
+                return Err(ErrorCode::UnknownException(
+                    "Data mutates during operation".to_string(),
+                ));
+            }
+
+            // Read all segments information in parallel.
+            let fuse_segment_io = SegmentsIO::create(ctx.clone(), self.operator.clone());
+            let results = fuse_segment_io.read_segments(&new_segments).await?;
+            for result in results.iter() {
+                let segment = result.clone()?;
+                summary = merge_statistics(&summary, &segment.summary)?;
+            }
+            new_segments.extend(segments.clone());
+            segments = new_segments;
+        }
+
+        let mut new_snapshot = TableSnapshot::from_previous(&latest_snapshot);
+        new_snapshot.segments = segments;
+        new_snapshot.summary = summary;
+
+        Self::commit_to_meta_server(
```

Comment:
shall we retry the commit if version-mismatched error is detected, and the refreshed snapshot is appended only?
there are cases the same table is kept being appended concurrently at a not-that-low frequency.